tests/async_tests/test_retries.py

```diff
@@ -0,0 +1,152 @@
+import queue
+import time
+from typing import Any, List, Optional
+
+import pytest
+
+import httpcore
+from httpcore._backends.auto import AsyncSocketStream, AutoBackend
+from tests.utils import Server
+
+
+class AsyncMockBackend(AutoBackend):
+    def __init__(self) -> None:
+        super().__init__()
+        self._exceptions: queue.Queue[Optional[Exception]] = queue.Queue()
+        self._timestamps: List[float] = []
+
+    def push(self, *exceptions: Optional[Exception]) -> None:
+        for exc in exceptions:
+            self._exceptions.put(exc)
+
+    def pop_open_tcp_stream_intervals(self) -> list:
+        intervals = [b - a for a, b in zip(self._timestamps, self._timestamps[1:])]
+        self._timestamps.clear()
+        return intervals
+
+    async def open_tcp_stream(self, *args: Any, **kwargs: Any) -> AsyncSocketStream:
+        self._timestamps.append(time.time())
+        exc = None if self._exceptions.empty() else self._exceptions.get_nowait()
+        if exc is not None:
+            raise exc
+        return await super().open_tcp_stream(*args, **kwargs)
+
+
+async def read_body(stream: httpcore.AsyncByteStream) -> bytes:
+    try:
+        body = []
+        async for chunk in stream:
+            body.append(chunk)
+        return b"".join(body)
```

Comment:
Nit: I see this `body.append(chunk)` a lot when I think the below is much cleaner:
```suggestion
        return b"".join([chunk async for chunk in stream])
```

Reply:
Yes, fair enough, just copy-pasted some code from other modules. :-)